lib/std/ihex.zig

```diff
@@ -0,0 +1,255 @@
+                checksum +%= std.fmt.parseInt(u8, line[i .. i + 2], 16) catch return error.InvalidRecord;
+            }
+            if (checksum != 0)
+                return error.InvalidChecksum;
+        }
+
+        switch (record_type) {
+            0x00 => {
+                var temp_data: [255]u8 = undefined;
+                for (temp_data[0..byte_count]) |*c, i| {
+                    c.* = std.fmt.parseInt(u8, line[8 + 2 * i .. 10 + 2 * i], 16) catch return error.InvalidRecord;
+                }
+                const data = Record.Data{
+                    .offset = address,
+                    .data = temp_data[0..byte_count],
+                };
+
+                try loader(context, Record{ .data = data });
+            },
+            0x01 => {
+                try loader(context, Record{ .end_of_file = {} });
+                return;
+            },
+            0x02 => {
+                if (byte_count != 2)
+                    return error.InvalidRecord;
+                const addr = Record.ExtendedSegmentAddress{
+                    .segment = std.fmt.parseInt(u16, line[8..12], 16) catch return error.InvalidRecord,
+                };
+                try loader(context, Record{ .extended_segment_address = addr });
+            },
+            0x03 => {
+                if (byte_count != 4)
+                    return error.InvalidRecord;
+                const addr = Record.StartSegmentAddress{
+                    .segment = std.fmt.parseInt(u16, line[8..12], 16) catch return error.InvalidRecord,
+                    .offset = std.fmt.parseInt(u16, line[12..16], 16) catch return error.InvalidRecord,
+                };
+                try loader(context, Record{ .start_segment_address = addr });
+            },
+            0x04 => {
+                if (byte_count != 2)
+                    return error.InvalidRecord;
+                const addr = Record.ExtendedLinearAddress{
+                    .upperWord = std.fmt.parseInt(u16, line[8..12], 16) catch return error.InvalidRecord,
+                };
+                try loader(context, Record{ .extended_linear_address = addr });
+            },
+            0x05 => {
+                if (byte_count != 4)
+                    return error.InvalidRecord;
+                const addr = Record.LinearStartAddress{
+                    .offset = std.fmt.parseInt(u32, line[8..16], 16) catch return error.InvalidRecord,
+                };
+                try loader(context, Record{ .start_linear_address = addr });
+            },
+            else => return error.InvalidRecord,
+        }
+    }
+}
+
+/// Parses intel hex data segments from the stream, using `mode` as parser configuration.
+/// For each data record, `loader` is called with `context` as the first parameter.
+pub fn parseData(stream: anytype, mode: ParseMode, context: anytype, Errors: type, loader: fn (@TypeOf(context), offset: u32, record: []const u8) Errors!void) !?u32 {
```

Comment:
Using callbacks like this is a smell. This inverts the control flow, making the code harder to read for the human, and  messes up the compiler's ability to create a call graph. I'm getting the impression from this pull request that IHEX is so simple that it doesn't really make sense to have a general-purpose abstraction, and `std.ihex` may be better suited to be a "bits" file, for example like dwarf_bits.zig or most of elf.zig (minus the higher level functions).